There's still this warning on build:

```
/usr/src/openwrt/build_dir/target-x86_64_musl/linux-x86_64/beadapanel-1.0-1/beada.c:97:61: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
   97 |                         sprintf(binstr, "%08x -", i + (int)addr);
```

--- a/src/beada.c
+++ b/src/beada.c
@@ -56,7 +56,7 @@
 #define RGB565_BPP			16
 #define CMD_TIMEOUT			msecs_to_jiffies(200)
 #define DATA_TIMEOUT			msecs_to_jiffies(1000)
-#define PANELLINK_MAX_DELAY		msecs_to_jiffies(2000)
+#define PANELLINK_MAX_DELAY		msecs_to_jiffies(2500)
 #define CMD_SIZE			512*4
 
 struct beada_device {

--- a/src/beada.c
+++ b/src/beada.c
@@ -59,6 +59,10 @@
 #define PANELLINK_MAX_DELAY		msecs_to_jiffies(2500)
 #define CMD_SIZE			512*4
 
+static int landscape = 0;
+module_param(landscape, int, 0660);
+MODULE_PARM_DESC(landscape, "enable landscape mode by setting this value to 1");
+
 struct beada_device {
 	struct drm_device				dev;
 	struct drm_simple_display_pipe	pipe;
@@ -84,6 +88,8 @@ struct beada_device {
 	unsigned int	misc_rcv_ept;
 	unsigned int	misc_snd_ept;
 	unsigned int	data_snd_ept;
+
+	bool		landscape;
 };
 
 #define to_beada(__dev) container_of(__dev, struct beada_device, dev)
@@ -294,11 +300,11 @@ static int beada_misc_request(struct beada_device *beada)
 		break;
 	}
 
-	beada->width = width;
-	beada->height = height;
+	beada->width = beada->landscape ? height : width;
+	beada->height = beada->landscape ? width : height;
 	beada->model = model;
-	beada->width_mm = width_mm;
-	beada->height_mm = height_mm;
+	beada->width_mm = beada->landscape ? height_mm : width_mm;
+	beada->height_mm = beada->landscape ? width_mm : height_mm;
 	
 	return 0;
 }
@@ -348,6 +354,7 @@ static void beada_fb_mark_dirty(struct drm_framebuffer *fb, const struct iosys_m
 		(beada->old_rect.x2 == 0) &&
 		(beada->old_rect.y2 == 0)) ) {
 
+		//snprintf(fmtstr, sizeof(fmtstr), "video/x-raw, format=RGB16, height=%d, width=%d, framerate=0/1", height, width);
 		snprintf(fmtstr, sizeof(fmtstr), "image/x-raw, format=BGR16, height=%d, width=%d, framerate=0/1", height, width);
 		ret = beada_send_tag(beada, (const char *)fmtstr);
 		if (ret < 0)
@@ -469,6 +476,7 @@ static void beada_pipe_enable(struct drm_simple_display_pipe *pipe,
 				 struct drm_crtc_state *crtc_state,
 				 struct drm_plane_state *plane_state)
 {
+	struct beada_device *beada = to_beada(fb->dev);
 	struct drm_shadow_plane_state *shadow_plane_state = to_drm_shadow_plane_state(plane_state);
 	struct drm_framebuffer *fb = plane_state->fb;
 	struct drm_rect rect = {
@@ -477,6 +485,8 @@ static void beada_pipe_enable(struct drm_simple_display_pipe *pipe,
 		.y1 = 0,
 		.y2 = fb->height,
 	};
+
+	pipe->crtc.rotation = beada->landscape ? DRM_MODE_ROTATE_90 : DRM_MODE_ROTATE_0;
 
 	beada_fb_mark_dirty(fb, &shadow_plane_state->data[0], &rect);
 }
@@ -620,6 +630,54 @@ static void beada_edid_setup(struct beada_device *beada)
 	beada_edid.checksum = beada_edid_block_checksum((u8*)&beada_edid);
 }
 
+static ssize_t landscape_get(struct device* dev, struct device_attribute *attr, char *buf) {
+
+	struct beada_device *beada = dev_get_drvdata(dev);
+
+	return sprintf(buf, beada->landscape ? "true\n" : "false\n");
+}
+
+static struct device_attribute landscape_attr = {
+	.attr = {
+		.name = "landscape",
+		.mode = S_IWUSR | S_IRUGO,
+	},
+	.show = landscape_get,
+};
+
+static ssize_t model_get(struct device* dev, struct device_attribute *attr, char *buf) {
+
+	struct beada_device *beada = dev_get_drvdata(dev);
+
+	return sprintf(buf, "BeadaPanel %s\n", beada->model);
+}
+
+static struct device_attribute model_attr = {
+	.attr = {
+		.name = "model",
+		.mode = S_IWUSR | S_IRUGO,
+	},
+	.show = model_get,
+};
+
+
+static int beada_sysfs_create(struct beada_device *beada)
+{
+	int ret;
+
+	ret = sysfs_create_file(&beada -> dev.dev -> kobj, &landscape_attr.attr);
+	if (ret)
+		return ret;
+
+	ret = sysfs_create_file(&beada -> dev.dev -> kobj, &model_attr.attr);
+	return ret;
+}
+
+static void beada_sysfs_remove(struct beada_device *beada)
+{
+	sysfs_remove_file(&beada -> dev.dev -> kobj, &landscape_attr.attr);
+	sysfs_remove_file(&beada -> dev.dev -> kobj, &model_attr.attr);
+}
 
 static int beada_usb_probe(struct usb_interface *interface,
 			  const struct usb_device_id *id)
@@ -642,6 +700,8 @@ static int beada_usb_probe(struct usb_interface *interface,
 		return PTR_ERR(beada);
 	}
 	
+	beada->landscape = landscape == 0 ? false : true;
+
 	/* Check corresponding endpoint number */
 	beada->misc_snd_ept = 2;
 	beada->misc_rcv_ept = 2;
@@ -718,7 +778,11 @@ static int beada_usb_probe(struct usb_interface *interface,
 
 	drm_fbdev_generic_setup(dev, 0);
 
-	DRM_DEV_INFO(&beada->udev->dev, "BeadaPanel %s detected\n", beada->model);
+	ret = beada_sysfs_create(beada);
+	if (ret)
+		DRM_DEV_ERROR(beada->dev.dev, "sysfs_create_file() return %d\n", ret);
+
+	DRM_DEV_INFO(&beada->udev->dev, "BeadaPanel %s detected, %s mode\n", beada->model, beada->landscape ? "landscape" : "portrait");
 
 	DRM_DEV_DEBUG(&beada->udev->dev, "--------------beada_usb_probe() exit\n");
 	return ret;
@@ -736,6 +800,8 @@ static void beada_usb_disconnect(struct usb_interface *interface)
 	struct beada_device *beada = to_beada(dev);
 
 	DRM_DEV_DEBUG(&beada->udev->dev, "--------------beada_usb_disconnect() enter\n");
+
+	beada_sysfs_remove(beada);
 
 	put_device(beada->dmadev);
 	beada->dmadev = NULL;

--- a/src/beada.c
+++ b/src/beada.c
@@ -94,7 +94,7 @@ void HexDump(unsigned char *buf, int len, unsigned char *addr) {
 
 	for (i = 0; i < len; i++) {
 		if (0 == (i % 16)) {
-			sprintf(binstr, "%08x -", i + (int)addr);
+			sprintf(binstr, "%08x -", i + (int)*(unsigned char *)addr);
 			sprintf(binstr, "%s %02x", binstr, (unsigned char)buf[i]);
 		}
 		else if (15 == (i % 16)) {

